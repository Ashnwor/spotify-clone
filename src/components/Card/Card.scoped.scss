@import '../../styles/color-palette';

.card {
  position: relative;
  display: flex;
  max-width: 194px;
  height: 251px;
  box-sizing: border-box;
  gap: 16px;
  flex-direction: column;
  padding: 20px 20px 16px;
  background: $color-black-4;
  border-radius: 8px;

  &:hover .fab {
    @media (min-width: 768px) {
      opacity: 1;
    }
  }

  @media (max-width: 768px) {
    width: 124px;
    height: 200px;
    gap: 2px;
    padding: 0;
    background: transparent;
    border-radius: 0;
  }
}

.fab {
  position: absolute;
  width: 40px;
  height: 40px;
  cursor: pointer;
  opacity: 0;

  @media (min-width: 768px) {
    right: 16px;
    bottom: 16px;
    background-color: $color-spotify-green;
    border-radius: 500px;

    &:hover {
      right: 14.6px;
      bottom: 14.6px;
      width: 42.4px;
      height: 42.4px;
    }
  }

  @media (max-width: 768px) {
    width: 100%;
    height: 100%;
    opacity: 0;
  }
}

.fab-visible {
  @media (min-width: 768px) {
    opacity: 1;
  }
}

.fab-icon {
  position: absolute;
  top: calc(50% - 16px / 2);
  left: calc(50% - 16px / 2);
  width: 16px;
  height: 16px;
  color: $color-white-1;

  @media (max-width: 768px) {
    display: none;
  }
}

.cover-container {
  position: relative;
  width: 100%;
  padding-bottom: 100%;
  background-color: #333;

  @media (min-width: 768px) {
    box-shadow: 0 10px 30px 0 rgba(0, 0, 0, 0.3), 0 1px 2px 0 rgba(0, 0, 0, 0.2);
  }
}

.cover {
  position: absolute;
  top: 0;
  left: 0;
  display: block;
  width: 100%;
  height: 100%;
}

.title {
  display: inline-block;
  overflow: hidden;
  max-width: 100%;
  color: $color-white-1;
  font-size: 16px;
  font-weight: 700;
  letter-spacing: normal;
  line-height: 24px;
  text-overflow: ellipsis;
  text-transform: none;
  white-space: nowrap;
}

.description {
  overflow: hidden;
  color: $color-gray-3;
  font-size: 11px;
  font-weight: 400;
  letter-spacing: normal;
  line-height: 16px;
  text-overflow: ellipsis;
  text-transform: none;
  white-space: nowrap;
}
