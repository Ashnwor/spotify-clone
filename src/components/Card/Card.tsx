import { useContext } from 'react';
import ImageFadeIn from 'react-image-fade-in';
import isURL from 'validator/lib/isURL';
import styles from './Card.module.scss';
import { PlayFill, PauseFill } from '../icons';
import CurrentSong from '../../context/CurrentSong';
import PlayingStatus from '../../context/PlayingStatus';
import Song from '../../types/Song';

interface Props {
  song: Song;
}

const Card: React.FC<Props> = ({ song }) => {
  const { currentSong, setCurrentSong } = useContext(CurrentSong);
  const { playing, setPlaying } = useContext(PlayingStatus);

  return (
    <div className={styles.card}>
      <div className={styles['cover-container']}>
        <ImageFadeIn
          className={styles.cover}
          src={
            song.cover && !isURL(song.cover)
              ? `${process.env.REACT_APP_BASE_URL}/${song.cover}`
              : song.cover
          }
        />
      </div>
      <div className={styles.info}>
        <div className={styles.title} title={song.name}>
          {song.name}
        </div>
        <div className={styles.description} title={song.artist}>
          {song.artist}
        </div>
      </div>
      <button
        className={`${styles.fab} ${
          currentSong?._id === song._id && playing ? styles['fab-visible'] : ''
        }`}
        tabIndex={0} // TODO
        aria-hidden="true"
        title="Play"
        onClick={() => {
          if (currentSong?._id === song._id && playing) {
            setPlaying?.(false);
          } else {
            setCurrentSong?.(song);
          }
        }}
      >
        {currentSong?._id === song._id && playing ? (
          <PauseFill className={styles['fab-icon']} />
        ) : (
          <PlayFill className={styles['fab-icon']} />
        )}
      </button>
    </div>
  );
};

export default Card;
