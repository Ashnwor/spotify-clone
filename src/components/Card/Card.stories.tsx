import Card from './Card';

export default {
  title: 'Card',
};

export const card: React.FC = () => {
  return (
    <div
      style={{
        display: 'grid',
        gridTemplateColumns: 'repeat(auto-fill, minmax(164px, 1fr))',
        columnGap: 48,
      }}
    >
      <Card
        song={{
          _id: 'story_id_1',
          name: 'Title',
          artist: 'Artist',
          cover: 'https://via.placeholder.com/150',
        }}
      />
      <Card
        song={{
          _id: 'story_id_1',
          name: 'Title',
          artist: 'Artist',
          cover: 'https://via.placeholder.com/150',
        }}
      />
      <Card
        song={{
          _id: 'story_id_1',
          name: 'Title',
          artist: 'Artist',
          cover: 'https://via.placeholder.com/150',
        }}
      />
    </div>
  );
};
