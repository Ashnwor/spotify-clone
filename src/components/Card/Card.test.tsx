import { shallow } from 'enzyme';
import Card from './Card';

it('runs correctly', () => {
  const wrapper = shallow(
    <Card
      song={{
        _id: 'test_id',
        artist: 'test_artist',
        cover: 'https://via.placeholder.com/1000',
        name: 'test_name',
      }}
    />,
  );
  expect(wrapper).toMatchSnapshot();
});
