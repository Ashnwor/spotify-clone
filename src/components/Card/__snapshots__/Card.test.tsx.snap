// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`runs correctly 1`] = `
<div
  className="card"
>
  <div
    className="cover-container"
  >
    <ImageFadeIn
      className="cover"
      loadAsBackgroundImage={false}
      opacityTransition={0.5}
      src="undefined/"
    />
  </div>
  <div
    className="info"
  >
    <div
      className="title-container"
    >
      <span
        className="title"
        title="test_name"
      >
        test_name
      </span>
    </div>
    <div
      className="description"
      title="test_artist"
    >
      test_artist
    </div>
  </div>
  <button
    aria-hidden="true"
    className="fab "
    onClick={[Function]}
    tabIndex={0}
    title="Play"
  >
    <SvgCardPlay
      className="fab-icon"
    />
  </button>
</div>
`;
